Plutarch/Verify.hs

```diff
@@ -0,0 +1,277 @@
+
+----------------------- PData instances -------------------------------------------------
+
+instance PTryFrom PData (PAsData PInteger) where
+  type PExcess PData (PAsData PInteger) = PInteger
+  ptryFrom opq = do
+    ver <- tcont $ plet (pasInt # opq)
+    pure $ (punsafeCoerce opq, ver)
+
+instance PTryFrom PData (PAsData PByteString) where
+  type PExcess PData (PAsData PByteString) = PByteString
+  ptryFrom opq = do
+    ver <- tcont $ plet (pasByteStr # opq)
+    pure $ (punsafeCoerce opq, ver)
+
+instance PTryFrom PData PData where
+  type PExcess PData PData = PUnit
+  ptryFrom opq = pure $ (opq, pcon PUnit)
+
+instance {-# OVERLAPPING #-} PTryFrom PData (PAsData (PBuiltinList PData)) where
+  type PExcess PData (PAsData (PBuiltinList PData)) = PBuiltinList PData
+  ptryFrom opq = do
+    ver <- tcont $ plet (pasList # opq)
+    pure $ (punsafeCoerce opq, ver)
+
+instance
+  {-# OVERLAPS #-}
+  ( PTryFrom PData (PAsData a)
+  , PTryFrom PData (PAsData b)
+  ) =>
+  PTryFrom PData (PAsData (PBuiltinMap a b))
+  where
+  type PExcess PData (PAsData (PBuiltinMap a b)) = PBuiltinList (PBuiltinPair (PAsData a) (PAsData b))
+  ptryFrom opq = do
+    verMap <- tcont $ plet (pasMap # opq)
+    -- I've not obtained a `PBuiltinList (PBuiltinPair a b)`
+    let verifyPair :: Term _ (PBuiltinPair PData PData :--> PBuiltinPair (PAsData a) (PAsData b))
+        verifyPair = plam $ \tup -> unTermCont $ do
+          (verfst, _) <- ptryFrom @PData @(PAsData a) $ pfstBuiltin # tup
+          (versnd, _) <- ptryFrom @PData @(PAsData b) $ psndBuiltin # tup
+          pure $ ppairDataBuiltin # verfst # versnd
+    ver <- tcont $ plet $ pmap # verifyPair # verMap
+    pure (punsafeCoerce opq, ver)
+
+instance {-# OVERLAPPING #-} PTryFrom PData (PAsData (PBuiltinMap POpaque POpaque)) where
+  type PExcess PData (PAsData (PBuiltinMap POpaque POpaque)) = PBuiltinList (PBuiltinPair (PAsData POpaque) (PAsData POpaque))
+  ptryFrom opq = do
+    ver <- tcont $ plet (pasMap # opq)
+    pure $ (punsafeCoerce opq, punsafeCoerce ver) -- PAsData POpaque ~ PData
+
+instance
+  {-# OVERLAPPABLE #-}
+  ( PTryFrom PData a
+  , a ~ PAsData b
+  , PIsData b
+  ) =>
+  PTryFrom PData (PAsData (PBuiltinList a))
+  where
+  type PExcess PData (PAsData (PBuiltinList a)) = PBuiltinList a
+  ptryFrom opq =
+    let lst :: Term _ (PBuiltinList a)
+        lst = punsafeBuiltin PLC.UnListData # opq
+        verify :: a ~ PAsData b => Term _ (PAsData b :--> a)
+        verify = plam $ \e ->
+          unTermCont $ do
+            (wrapped, _) <- ptryFrom @PData @a $ pforgetData e
+            pure wrapped
+     in do
+          ver <- tcont $ plet $ pmap # verify # lst
+          pure $ (punsafeCoerce opq, ver)
+
+instance
+  {-# OVERLAPPABLE #-}
+  ( PTryFrom PData a
+  , a ~ PAsData a'
+  , PIsData a'
+  , PTryFrom PData b
+  , b ~ PAsData b'
+  , PIsData b'
+  ) =>
+  PTryFrom PData (PAsData (PBuiltinPair a b))
```

Comment:
it's better to have a verifier for `PTuple` rather than `PAsData (PBuiltinPair ... )` - they are both the same and no op conversions exist. The former is certainly easier to work with without overlapping instances and stuff though.

Reply:
I just marked it as `OVERLAPPABLE` because there will be another instance for `PBuiltinPair PData PData`, as well.

Reply:
`PTuple` should automatically be an instance of `PTryFrom` because it's just a type alias for a `PDataSum`.